ddclient: T5612: Miscellaneous improvements and fixes for dynamic DNS

indrajitr opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
- Fix VRF support for ddclient service
- Improve dual stack support for dyndns2 protocol
- Generate more reliable ddclient config
- Refactor zone configuration
- Enable TTL support for web-service based protocols
- Adjust validator and completion for ddclient
- Additional cleanup and refactoring for ddclient scripts
- Relax hostname validation for apex and wildcard entry

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://vyos.dev/T5612

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
dns dynamic

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
```
set service dns dynamic address eth5 service cloudflare host-name 'dyn.example.com'
set service dns dynamic address eth5 service cloudflare ip-version 'both'
set service dns dynamic address eth5 service cloudflare password 'scrambled_string'
set service dns dynamic address eth5 service cloudflare protocol 'cloudflare'
set service dns dynamic address eth5 service cloudflare ttl '300'
set service dns dynamic address eth5 service cloudflare zone 'example.com'

set service dns dynamic address eth5 service namecheap host-name '@.example.com'
set service dns dynamic address eth5 service namecheap username 'example.com'
set service dns dynamic address eth5 service namecheap password 'scrambled_string'
set service dns dynamic address eth5 service namecheap protocol 'namecheap'

```

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
vyos@vyos15:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_dns_dynamic.py 
test_01_dyndns_service_standard (__main__.TestServiceDDNS.test_01_dyndns_service_standard) ... 

"freedns" does not support "zone"


"freedns" does not support "ttl"


"zoneedit1" does not support "zone"


"zoneedit1" does not support "ttl"

ok
test_02_dyndns_service_ipv6 (__main__.TestServiceDDNS.test_02_dyndns_service_ipv6) ... ok
test_03_dyndns_service_dual_stack (__main__.TestServiceDDNS.test_03_dyndns_service_dual_stack) ... 
"googledomains" does not support both IPv4 and IPv6 at the same time

ok
test_04_dyndns_rfc2136 (__main__.TestServiceDDNS.test_04_dyndns_rfc2136) ... ok
test_05_dyndns_hostname (__main__.TestServiceDDNS.test_05_dyndns_hostname) ... ok
test_06_dyndns_vrf (__main__.TestServiceDDNS.test_06_dyndns_vrf) ... ok

----------------------------------------------------------------------
Ran 6 tests in 220.985s

OK
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

indrajitr committed:
ddclient: T5612:  Fix VRF support for ddclient service

Fix VRF support interface definition and configuration mode for ddclient
to actually capture the VRF name and pass it to the template.

indrajitr committed:
ddclient: T5612: Improve dual stack support for dyndns2 protocol

dyndns2 protocol in ddclient honors dual stack for selective servers
because of the way it is implemented in ddclient.

We formalize the well known servers that support dual stack in a list
and check against it when validating the configuration.

indrajitr committed:
ddclient: T5612: Generate more reliable ddclient config

Adjust the jinja template to avoid generating incorrect ddclient.conf in
some cases. The template is reformatted to guarantee whitespacing and
empty line separation.

indrajitr committed:
ddclient: T5612: Additional refactoring for scripts and smoketests

Additional cleanup and refactoring for ddclient scripts including the
smotektests.

indrajitr committed:
ddclient: T5612: Enable TTL support for web-service based protocols

Enable TTL support for web-service based protocols in addition to
RFC2136 based (nsupdate) protocol.

Since TTL is not supported by all protocols, and thus cannot have a
configuration default, the existing XML snippet `include/dns/time-to-live.xml.i`
does not have common `<defaultValue>300</defaultValue>` anymore and is
instead added explicitly whenever necessary.

indrajitr committed:
ddclient: T5612: Refactor zone configuration

Refactor zone configuration to use shared XML snippet for all cases.

indrajitr committed:
ddclient: T5612: Adjust validator and completion for ddclient

Adjust the validator and completion for ddclient to remove unsupported
or superfluous protocols.

Specifically,
- remove 'nsupdate' protocol from the list because there is a separate
  config path for that protocol (rfc2136)
- remove 'cloudns' protocol from the list because it has non standard
  configuration and is not supported by our configurator at this time

c-po reviewed (changes requested):
VRF support is a very nice addition, but it will only work with one VRF due to the current design (only one global ddclient process) - still better then nothing.

Can you please extend the smoketests to also verify the VRF case and check if ddclient process runs in the VRF it was assigned too?

indrajitr committed:
ddclient: T5612: Relax hostname validation for apex and wildcard entry

Some porvides (like 'namecheap') allow to use '@' or '*' as hostname
prefix for apex and wildcard records. This commit relaxes the hostname
validation to allow these prefixes.

indrajitr commented:
@c-po Thank you for the review!
> VRF support is a very nice addition, but it will only work with one VRF due to the current design (only one global ddclient process) - still better then nothing.

Yes, maybe we can consider templated service at some point.

> Can you please extend the smoketests to also verify the VRF case and check if ddclient process runs in the VRF it was assigned too?

Have updated and expanded smoketests. They cover VRF now.

sever-sever commented:
@MergifyIo backport sagitta
